### PR TITLE
changes to driver page

### PR DESCRIPTION
**Pull request by K843Barber**
Made some additions to the driver pages. You may want to tidy it a little, but got the results aside the Fastest Laps. Still to figure out how to skip "Withdrew" comments

### CODE REVIEW

File: src/pages/drivers/[driver].tsx [hunk 1]
```diff
@@ -70,7 +70,31 @@ const individualDriverSchema = z.object({
...
+  // var wins = 0
+  // for (let i = 0; i < resultsQuery.data.length; i++)
+  // {
+  //   if (resultsQuery.data[i].Results[0].position === "1")
+  //   {
+  //     wins += 1
+  //   }
+  // }
   return
+  // return wins
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: This is commented out, but for reference: you're using a [react-query hook `useQueries`](https://tanstack.com/query/v4/docs/react/reference/useQueries) and a [NextJS hook `useRouter`](https://nextjs.org/docs/api-reference/next/router) inside a normal function (i.e. not a React component), so it won't work. A React component using hooks either returns [JSX](https://facebook.github.io/jsx/) or the return value of a [custom hook](https://reactjs.org/docs/hooks-custom.html)

File: src/pages/drivers/[driver].tsx [hunk 2]
```diff
@@ -98,16 +122,51 @@ export default function Driver() {
...
+    var FastestLaps = 0
+
+    for (let i = 0; i < resultsQuery.data.length; i++)
+    {
+      if (resultsQuery.data[i].Results[0].position === "1"){wins += 1}
+      if (resultsQuery.data[i].Results[0].grid === "1"){poles += 1}
+      // if (resultsQuery.data[i].Results[0].status !== "Withdrew")
+      // {
+      //   if (resultsQuery.data[i].Results[0].FastestLap.rank === "1"){FastestLaps += 1}
+      // }
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: I would probably move this for loop out of the component and create a custom function. You're accessing similar things from the API, so you could just create one function that returns whatever you pass to it. 

I'd first create a schema based on the data we want to get: 

```ts
const resultsSchema = z.object({
  Constructor: z.object({
    constructorId: z.string(),
    name: z.string(),
    nationality: z.string(),
  }),
  Driver: z.object({
    code: z.string(),
    dateOfBirth: z.string(),
    driverId: z.string(),
    familyName: z.string(),
    givenName: z.string(),
    nationality: z.string(),
    permanentNumber: z.string(),
  }),
  grid: z.string(),
  laps: z.string(),
  number: z.string(),
  points: z.string(),
  position: z.string(),
  positionText: z.string(),
  status: z.string(),
})
```

Then I'd narrow in on that and create a races schema to get the specific data we're interested in: 
```ts
const racesSchema = z.object({
  Circuit: z.object({
    Location: z.object({
      country: z.string(),
      lat: z.string(),
      locality: z.string(),
      long: z.string(),
    }),
  }),
  Results: z.array(resultsSchema),
  date: z.string(),
  raceName: z.string(),
  round: z.string(),
  season: z.string(),
})
```
and create a TypeScript type from it: 

```ts
type Races = z.infer<typeof racesSchema>
```

I'd then create a helper function that essentially just adds up the amount of things you want: 

```ts
type Kind = 'position' | 'grid'
function getDriverStats(
  results: Races[],
  key: Kind,
  value: number | string
) {
  return results?.reduce((total, current) => {
    if (current?.Results[0][key] === value) {
      total += 1
    }
    return total
  }, 0)
}
```

Then, in the Driver component, you can access these values like so: 
```tsx
const wins = getDriverStats(
  resultsQuery.data,
  'position',
  '1'
)
const poles = getDriverStats(
  resultsQuery.data,
  'grid',
  '1'
)
```

**Comment by K843Barber** [comment]
Made the changes to what you mention above

File: src/pages/drivers/[driver].tsx [hunk 3]
```diff
@@ -69,10 +69,105 @@ const individualDriverSchema = z.object({
...
+) {
+  return results?.reduce((total, current) => {
+    if (current?.Results[0][key] === value) {
+      total += 1
+    }
+    return total
+  }, 0)
 }
 
+type Races1 = z.infer<typeof racesSchema>
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: You don't need to make a new type here as it's the same as the `Races` type. You can just use `Races`.

File: src/pages/drivers/[driver].tsx [hunk 4]
```diff
@@ -69,10 +69,105 @@ const individualDriverSchema = z.object({
...
+type Kind1 = 'rank'
+type Kind2 = 'points'
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: I would just add these two options to the `Kind` union, like: 

`type Kind = 'position' | 'grid' | 'rank' | 'points'`

File: src/pages/drivers/[driver].tsx [hunk 5]
```diff
@@ -69,10 +69,105 @@ const individualDriverSchema = z.object({
...
+
+function getDriverStats1(
+  results: Races1[],
+  key: Kind1,
+  value: number | string
+) {
+  return results?.reduce((total, current) => {
+    if (current?.Results[0]['status'] != "Withdrew"){
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: Totally up to you, but you could just use dot notation for consistency

```suggestion
    if (current?.Results[0].status != "Withdrew") {
```

File: src/pages/drivers/[driver].tsx [hunk 6]
```diff
@@ -99,15 +194,59 @@ export default function Driver() {
       </p>
     )
 
-  console.log(driverQuery.data)
-  console.log(resultsQuery.data)
+    const wins = getDriverStats(
+      resultsQuery.data,
```
Comment (nialldbarber) [inline src/pages/drivers/[driver].tsx]: Given that `resultsQuery.data` is used a few times, you could consider destructuring it: 
```js
const { data } = resultsQuery
```

File: src/pages/drivers/[driver].tsx (hunk 1 above)
Comment (K843Barber) [reply src/pages/drivers/[driver].tsx]: Done

File: src/pages/drivers/[driver].tsx (hunk 2 above)
Comment (K843Barber) [reply src/pages/drivers/[driver].tsx]: Done

File: src/pages/drivers/[driver].tsx (hunk 3 above)
Comment (K843Barber) [reply src/pages/drivers/[driver].tsx]: Done

File: src/pages/drivers/[driver].tsx (hunk 4 above)
Comment (K843Barber) [reply src/pages/drivers/[driver].tsx]: Done

File: src/pages/drivers/[driver].tsx (hunk 5 above)
Comment (K843Barber) [reply src/pages/drivers/[driver].tsx]: Not Done

**Pull request merged**